crates/wasi-provider/src/wasi_runtime.rs

```diff
@@ -71,64 +88,246 @@ impl WasiRuntime {
-                let export = match i.module() {
-                    "wasi_snapshot_preview1" => wasi_snapshot.get_export(i.name()),
-                    "wasi_unstable" => wasi_unstable.get_export(i.name()),
-                    other => bail!("import module `{}` was not found", other),
-                };
-                match export {
-                    Some(export) => Ok(export.clone().into()),
-                    None => bail!(
-                        "import `{}` was not found in module `{}`",
-                        i.name(),
-                        i.module()
-                    ),
+        // Clone the module data so it can be moved and we don't have to worry
+        // about the lifetime of the struct data
+        let module_data = self.module_data.clone();
+        // We could get multiple status updates, so this gives a little
+        // breathing room while avoiding blocking. This is currently super
+        // naive, but will work for now. In the future, we may just want to use
+        // a try_send with a retry
+        let (status_sender, status_recv) = watch::channel(ContainerState {
+            waiting: Some(ContainerStateWaiting {
+                message: Some("No status has been received from the process".into()),
+                reason: None,
+            }),
+            ..Default::default()
+        });
+        let handle = tokio::task::spawn_blocking(move || -> Result<_, failure::Error> {
+            let engine = wasmtime::Engine::default();
+            let store = wasmtime::Store::new(&engine);
+            let wasi_snapshot = Wasi::new(&store, wasi_ctx_snapshot);
+            let wasi_unstable = WasiUnstable::new(&store, wasi_ctx_unstable);
+            let module = match wasmtime::Module::new(&store, &module_data) {
+                // We can't map errors here or it moves the send channel, so we
+                // do it in a match
+                Ok(m) => Ok(m),
+                Err(e) => {
+                    let message = "unable to create module";
+                    error!("{}: {:?}", message, e);
+                    status_sender
+                        .broadcast(ContainerState {
+                            terminated: Some(ContainerStateTerminated {
+                                message: Some(message.into()),
+                                reason: None,
+                                exit_code: 1,
+                                finished_at: Some(Time(chrono::Utc::now())),
+                                ..Default::default()
+                            }),
+                            ..Default::default()
+                        })
+                        .expect("status should be able to send");
+                    // Converting from anyhow
+                    Err(format_err!("{}: {}", message, e))
+                }
+            }?;
+            // Iterate through the module includes and resolve imports
+            let imports = match module
```

Comment:
I think it would be easier to read if you separated the iteration of the imports from the match expression

Reply:
Normally I don't weigh in on style issues, but I agree with Ryan here: I found this difficult to read. There were 19 lines between the `match` and the beginning of the match block.

Reply:
Completely right. IT has been separated out